### PR TITLE
FDT Base Functionality

### PR DESCRIPTION
**Pull request by kjhayes**
Adds support for reading the FDT, traversing nodes and properties

Has some additional helper  functions to deal with really common use cases like searching for nodes based on compatiblity

Need to handle the memresv map in the FDT still though, but other than that this should suffice for now (and possibly forever if we can use this to get all the way to unflattening the tree)

### CODE REVIEW

File: arch/riscv/kernel.c [hunk 1]
```diff
@@ -74,10 +106,21 @@ kmain(uint64_t hartid, struct fdt* fdt) {
     putchar('\n');
 
     // Check the FDT
-    if (verify_fdt(fdt)) {
-        puts("Using a supported FDT Version\n");
+    if (fdt_verify(fdt)) {
+        puts("Cannot Read the FDT!\n");
+        // PANIC! (this is actually unrecoverable)
```
Comment (egelja) [inline arch/riscv/kernel.c]:
```suggestion
        abort();
```

File: include/kernel/stdint.h [hunk 2]
```diff
@@ -22,9 +22,11 @@ typedef unsigned long long uint64_t;
 #if CONFIG_64_BIT
 typedef uint64_t size_t;
 typedef int64_t ssize_t;
+typedef uint64_t uintptr_t;
 #else
 typedef uint32_t size_t;
 typedef int32_t ssize_t;
+typedef uint32_t uintptr_t;
```
Comment (egelja) [inline include/kernel/stdint.h]: this is done in the libc PR please rebase against it

File: include/kernel/string.h [hunk 3]
```diff
@@ -12,8 +12,8 @@ extern "C" {
 #endif
 
 void memcpy(uint8_t* src, uint8_t* dest, size_t len);
-int strlen(char*);
-int strcmp(char* a, char* b);
+int strlen(const char*);
+int strcmp(const char* a, const char* b);
```
Comment (egelja) [inline include/kernel/string.h]: this is done in the libc PR please rebase against it

File: toolchain/flags.cmake [hunk 4]
```diff
@@ -2,11 +2,11 @@
 
 set(
   CMAKE_C_FLAGS
-  "-ffreestanding -nostdlib -nostdinc -fdiagnostics-color=always -fno-common ${CMAKE_C_FLAGS}"
+  "-ffreestanding -nostdlib -nostdinc -fdiagnostics-color=always -fno-common -Wno-pointer-arith ${CMAKE_C_FLAGS}"
 )
 set(
   CMAKE_CXX_FLAGS
-  "-ffreestanding -nostdlib -nostdinc -nostdinc++ -fdiagnostics-color=always -fno-common -fno-rtti -fno-exceptions ${CMAKE_CXX_FLAGS}"
+  "-ffreestanding -nostdlib -nostdinc -nostdinc++ -fdiagnostics-color=always -fno-common -fno-rtti -fno-exceptions -Wno-pointer-arith ${CMAKE_CXX_FLAGS}"
```
Comment (egelja) [inline toolchain/flags.cmake]: no no no no no

this should go in `cmake/flags.cmake`

this is for base flags needed to get the toolchain to even work

File: kernel/string.c
Comment (egelja) [inline kernel/string.c:1]: this is done in the libc PR please rebase against it

File: kernel/of/fdt.c [hunk 5]
```diff
@@ -1,12 +1,340 @@
+
 #include <kernel/endian.h>
 #include <kernel/of/fdt.h>
 
+static int
```
Comment (egelja) [inline kernel/of/fdt.c]:
```suggestion
static bool
```

File: kernel/of/fdt.c [hunk 6]
```diff
@@ -1,12 +1,340 @@
...
+            break;
+        case FDT_PROP:
+            prop = (struct fdt_prop*)token;
+            token++;
+            str_len = fdt_prop_val_len(prop);
+            cell_len = (str_len >> 2) + ((str_len & 3) != 0);
+            token += (2ull + cell_len);
+            break;
+        case FDT_END:
+        default:
```
Comment (egelja) [inline kernel/of/fdt.c]: this needs more comments

File: kernel/of/fdt.c [hunk 7]
```diff
@@ -1,12 +1,340 @@
...
+            token = NULL;
+            break;
+    }
+
+    if (fdt_struct_ptr_check(fdt, (void*)token)) {
+        return NULL;
```
Comment (egelja) [inline kernel/of/fdt.c:58]:
```suggestion
        return nullptr;
```

File: kernel/of/fdt.c [hunk 8]
```diff
@@ -1,12 +1,340 @@
...
+    return token;
+}
+
+static uint32_t*
+fdt_next_token_after_node(struct fdt* fdt, struct fdt_node* node) {
+    uint32_t* token_ptr = (uint32_t*)node;
+    if (node == NULL) {
+        return NULL;
+    }
+    int depth = 0;
```
Comment (egelja) [inline kernel/of/fdt.c:69]:
```suggestion
    size_t depth = 0;
```

File: kernel/of/fdt.c [hunk 9]
```diff
@@ -1,12 +1,340 @@
...
+        if (token_ptr == NULL) {
+            return NULL;
+        }
+
+    } while (depth > 0);
+
+    return token_ptr;
+}
+
 int
```
Comment (egelja) [inline kernel/of/fdt.c]:
```suggestion
bool
```

File: kernel/of/fdt.c [hunk 10]
```diff
@@ -1,12 +1,340 @@
...
-verify_fdt(struct fdt* fdt) {
-    return (be32toh(fdt->last_comp_version) < FDT_COMPAT_VERSION);
+fdt_verify(struct fdt* fdt) {
+    if (be32toh(fdt->magic) != 0xd00dfeed) {
```
Comment (egelja) [inline kernel/of/fdt.c]: shouldn't this be a define

File: kernel/of/fdt.c [hunk 11]
```diff
@@ -1,12 +1,340 @@
...
+        const char* prop_name = fdt_prop_name(fdt, prop);
+        if (prop_name != NULL && strcmp(name, prop_name) == 0) {
+            return prop;
+        }
+        prop = fdt_node_next_prop(fdt, prop);
+    }
+    return NULL;
+}
+
+int
```
Comment (egelja) [inline kernel/of/fdt.c]:
```suggestion
bool
```

**Review by egelja — CHANGES_REQUESTED** [review]
this is very good code

**Comment by egelja** [comment]
this rebase will be fun

**Comment by echavemann** [comment]
rebase not too bad this PR is hilarious tho

File: arch/riscv/kernel.c (hunk 1 above)
Comment (kjhayes) [reply arch/riscv/kernel.c]: abort() implies it's a user-space program to me which ig is fiiiine but I don't like it, can we define a kernel_panic() function or similar?

Comment (egelja) [reply arch/riscv/kernel.c]:
```cpp
#ifdef KERNEL
#define kernel_panic() abort()
#endif
```

put that into `libc/include/stdlib.h`

File: kernel/of/fdt.c (hunk 8 above)
Comment (kjhayes) [reply kernel/of/fdt.c:69]: depth can be negative in an ill formed tree (I know we check against >0 not >=0 but there's no good reason to make it unsigned or larger than an int)

File: kernel/of/fdt.c (hunk 10 above)
Comment (kjhayes) [reply kernel/of/fdt.c]: it could be, literally no other use case for it though

File: kernel/of/fdt.c (hunk 7 above)
Comment (kjhayes) [reply kernel/of/fdt.c:58]: Here there's literally no difference between NULL and nullptr afaik

Comment (egelja) [reply kernel/of/fdt.c:58]: ones a builtin. It's just "more right"

File: kernel/string.c
Comment (egelja) [inline kernel/string.c:1]: this was deleted

**Review by egelja — CHANGES_REQUESTED** [review]

File: kernel/string.c
Comment (egelja) [reply kernel/string.c:1]: in the libc pr

**Review by egelja — APPROVED** [review]
lgtm

**Review by egelja — APPROVED** [review]

**Pull request merged**